agent_build/docker_images/image_builder.py

```diff
@@ -0,0 +1,195 @@
+import hashlib
+import logging
+import pathlib as pl
+import subprocess
+from typing import Union, Dict, Optional
+
+from agent_build import common
+
+logging.basicConfig(level=logging.INFO)
+
+
+SOURCE_ROOT = pl.Path(__file__).absolute().parent.parent
+
+
+class ImageBuilder:
+    """
+    Simple abstraction to build docker image. It also has the additional methods which should be useful in the CI/CD to
+    cache the image for future use.
+    """
+
+    def __init__(
+            self, image_name: str,
+            source_dir: Union[str, pl.Path],
+            dockerfile_path: Union[str, pl.Path],
+            build_args: Dict[str, str] = None
+    ):
+        """
+        :param image_name: Name of the result image.
+        :param source_dir: The path for the image build context.
+        :param dockerfile_path: Path to the Dockerfile.
+        """
+        self._source_dir = pl.Path(source_dir)
+        self._dockerfile_path = pl.Path(dockerfile_path)
+        self._build_args = build_args or {}
+
+        self._image_name_prefix = image_name
+        self._image_name: Optional[str] = None
+
+        self._is_built = False
+
+    @property
+    def image_name(self):
+        if not self._image_name:
+            used_files_checksum_sha256 = common.get_files_sha256_checksum(self.get_used_files())
+            used_files_checksum = used_files_checksum_sha256.hexdigest()
+            self._image_name = f"{self._image_name_prefix}-{used_files_checksum}"
+
+        return self._image_name
+
+    def get_used_files(self):
```

Comment:
Do we already have some tests for ``get_used_files()`` function?

Reply:
This function is removed now. Things should be simpler now, I hope.